src/api/index.js

```diff
@@ -1,26 +1,38 @@
 import { TOP_FAVORITES } from '../constants/categories'
 import { getJSON } from './get'
 
-export function getFavorites({ eventType, pageSize, sortOrder }) {
-    const queryParams = `eventType=${eventType}&pageSize=${pageSize}&sortOrder=${sortOrder}&_=${Date.now()}`
-    return getJSON(`dataStatistics/favorites?${queryParams}`)
-}
+const api = (() => {
```

Comment:
Would there be another way to address this? Or would that be very involved? I'm not a big fan of passing around a stateful module. Something stateless (even if it's slightly more involved) would have my preference because it's usually more predictable and easier to reason about.

Reply:
Since we're using hooks anyway, we could just have a useApi hook for example.

Reply:
If that's out of scope or we're in a rush, I think it'd be a good follow up issue.

Reply:
I had this implemented as a class before, but thought the IIFE was a bit cleaner and not mixing patterns (functional vs OO) too much.... Ultimately, the problem is this: the functions that fetch from the server need to know a `baseUrl`. The most obvious way to get this baseUrl is using the `useConfig` hook. But you can only call hooks from within components or other hooks. So, without completely refactoring this app to use the dataEngine, and moving away from state management in redux, it seemed most logical to me that the api was something with an internal reference to a baseUrl, which can be set when the App renders....

Technically I think I could even just put everything that's in the IIFE directly in `./src/api/index.js`, but to me it then becomes less clear how all of this works together...

There is also issue [DHIS2-8514](https://jira.dhis2.org/browse/DHIS2-8514), which would refactor this app to use the DataEngine, and remove the api and redux related code. Since I knew that's coming, I thought I'd go for the simplest solution for now, and the IIFE was what I found simplest in the end.

If you have suggestions for a simple, low-effort, solution that would be cleaner, I'd be happy to implement it.

Reply:
I see, yeah that sounds good. Outside of more involved refactors I then also wouldn't know how to address this cleanly. In that case DHIS2-8514 sounds like the follow up where we'd address this. Thanks for explaining 👍

Reply:
@ismay @HendrikThePendric should we add `withConfig` as a HoC that can wrap class components?  I prefer the hooks approach, but maybe it's too restrictive.

Reply:
I personally prefer a hook over a HOC as well. I also think the issue here will be resolved once this app moves to the app platform:

> There is also issue DHIS2-8514, which would refactor this app to use the DataEngine, and remove the api and redux related code. Since I knew that's coming, I thought I'd go for the simplest solution for now, and the IIFE was what I found simplest in the end.

Reply:
A HOC wouldn't really solve the issue here either as I need the need the baseUrl somewhere fully outside of the React components. As @ismay says, this app needs to be refactored to use the app-platform's DataEngine (and get rid of redux) and that will resolve this issue.